ukat/qa/snr.py

```diff
@@ -2,6 +2,76 @@
 import numpy as np
 import os
 
+from sklearn.mixture import BayesianGaussianMixture
+
+
+class Isnr:
+    """
+        Attributes
+        ----------
+        isnr : float
+            Image signal to noise ratio.
+        noise_mask : np.ndarray
+            Mask of the ROI used to measure noise levels. Should be voxels
+            outside the body.
+        clusters : np.ndarray
+            The labels assigned to each voxel during automatic background
+            segmentation. This can be useful for debugging.
+        """
+
+    def __init__(self, pixel_array, noise_mask=None, n_clusters=3):
+        """Initialise a image signal to noise ratio (iSNR) class instance.
+        Parameters
+        ----------
+        pixel_array : np.ndarray
+            Array of voxels over which iSNR should be calculated.
+        noise_mask : np.ndarray, optional
+            A binary voxel mask where voxels representing background i.e.
+            outside the body, are True. If no mask is supplied, one is
+            estimated using a Bayesian Gaussian mixture model to segment
+            background voxels.
+        n_clusters : int, optional, optional
+            When using the automatic background segmentation this is the
+            total number of componenets the image is segmented into. The
+            component with the lowest mean is assumed to be background.
+            Default 3.
+        """
+        self.pixel_array = pixel_array
+        self.n_clusters = n_clusters
+        self.shape = pixel_array.shape
+        self.dimensions = len(pixel_array.shape)
+
+        if noise_mask is None:
+            self.__mask_background__()
+        else:
+            self.noise_mask = noise_mask
+            self.clusters = np.nan
+
+        self.isnr = np.nan
+        self.__snr__()
+
+    def __mask_background__(self):
```

Comment:
My reason for keeping it as a "private" method is because its not been tested with every type of data under the sun so I'm not 100% confident it will work in all cases (however its better to have something, find its issues then fix them than have nothing at all though!). As there's some very explicit instructions to check things carefully in the tutorial with examples of how its being used I don't think this is too irresponsible. Moving it to a function in tools would mean people would be less likely to see the caveats and use it without realising/thinking about if its worked correctly. That I've never wanted a whole body mask/background mask other than calculating iSNR so makes sense to keep it bundled in the iSNR class. 

Either way I wouldn't do it as a copy as it means there's two places to change anything in future.